rust/psibase_macros/src/graphql_macro.rs

```diff
@@ -0,0 +1,195 @@
+        impl Query {
+            #queries
+        }
+    };
+
+    impl_query.into()
+}
+
+pub fn table_query_macro_impl(item: TokenStream) -> TokenStream {
+    let args = parse_macro_input!(item as Args);
+
+    let query_name = args.query_name;
+    let table = args.table;
+    let record = Ident::new(format!("{}Record", table).as_str(), table.span());
+    let key_fn = args.key_fn.ident;
+    let key_ty = args.key_fn.bounds.to_token_stream();
+
+    let query = quote! {
+        #[allow(clippy::too_many_arguments)]
+        async fn #query_name(
+            &self,
+            first: Option<i32>,
+            last: Option<i32>,
+            before: Option<String>,
+            after: Option<String>,
+            gt: Option<#key_ty>,
+            ge: Option<#key_ty>,
+            lt: Option<#key_ty>,
+            le: Option<#key_ty>,
+        ) -> async_graphql::Result<async_graphql::connection::Connection<psibase::RawKey, #record>> {
+            let table_idx = #table::new().#key_fn ();
+            psibase::TableQuery::new(table_idx)
+                .first(first)
+                .last(last)
+                .before(before)
+                .after(after)
+                .gt(gt.map(|k| k.into()).as_ref())
+                .ge(ge.map(|k| k.into()).as_ref())
+                .lt(lt.map(|k| k.into()).as_ref())
+                .le(le.map(|k| k.into()).as_ref())
+                .query()
+                .await
+        }
+    };
+
+    TokenStream::from(query)
+}
+
+pub fn table_query_subindex_macro_impl(item: TokenStream) -> TokenStream {
+    let args = parse_macro_input!(item as Args);
+
+    let query_name = args.query_name;
+    let table = args.table;
+    let record = Ident::new(format!("{}Record", table).as_str(), table.span());
+    let key_fn = args.key_fn.ident;
+    let subindex_params = args.subindex_params;
+
+    if subindex_params.is_empty() {
+        abort!(
+            query_name,
+            "subindex query must have at least one parameter"
+        );
+    }
+
+    let mut params = quote! {};
+    let mut subindex_args = quote! {};
+
+    // TODO: Should string be accepted as a default rest type of any subindex?
+    let mut subindex_rest_ty = quote! { String };
+
+    for param in subindex_params {
+        let param_name = param.ident;
+        let param_type = param.bounds.to_token_stream();
+        if param_name == "subindex_rest_ty" {
+            subindex_rest_ty = quote! { #param_type };
+        } else {
+            params = quote! {
+                #params
+                #param_name: #param_type,
+            };
+
+            subindex_args = quote! {
+                #subindex_args
+                &#param_name,
+            };
+        }
+    }
+
+    let query = quote! {
+        async fn #query_name(
+            &self,
+            #params
+            first: Option<i32>,
+            last: Option<i32>,
+            before: Option<String>,
+            after: Option<String>,
```

Comment:
Is there are reason that `gt`, `lt`, &c. are not used for subindex?

Reply:
Keeping these separate allows restricting the range with these args within a subindex

Reply:
Related: it'd be nice for table_query_subindex to support lt, gt, le, ge, but will require the macro user specifying a type for these. Maybe a future enhancement.

Reply:
Isn't that type `subindex_rest_ty`?

Reply:
> If the user wants to specify these arguments explicitly, I can't see how to do so using this macro
I meant having the macro automatically add them. Since `subindex_rest_ty` already exists (I missed that), it be extended to do it now.